src/Bicep.Core/FileSystem/FileResolver.cs

```diff
@@ -29,86 +29,65 @@ public FileResolver(IFileSystem fileSystem)
         }
 
         public ResultWithDiagnostic<string> TryRead(Uri fileUri)
+            => TryReadInternal<string>(fileUri, 0, stream => {
+                using var reader = new StreamReader(stream);
+
+                return new(reader.ReadToEnd());
+            });
+
+        public static ResultWithDiagnostic<FileWithEncoding> ReadWithEncoding(BinaryData data, Encoding fileEncoding, int maxCharacters, Uri fileUri)
         {
-            if (!fileUri.IsFile)
-            {
-                return new(x => x.UnableToLoadNonFileUri(fileUri));
-            }
+            using var sr = new StreamReader(data.ToStream(), fileEncoding, true);
 
-            try
+            Span<char> buffer = stackalloc char[LanguageConstants.MaxLiteralCharacterLimit + 1];
+            var sb = new StringBuilder();
+            while (!sr.EndOfStream)
             {
-                if (DirExists(fileUri))
+                var i = sr.ReadBlock(buffer);
+                sb.Append(new string(buffer.Slice(0, i)));
+                if (maxCharacters > 0 && sb.Length > maxCharacters)
                 {
-                    return new(x => x.FoundDirectoryInsteadOfFile(fileUri.LocalPath));
+                    return new(x => x.FileExceedsMaximumSize(fileUri.LocalPath, maxCharacters, "characters"));
                 }
-
-                ApplyWindowsConFileWorkaround(fileUri.LocalPath);
-                return new(fileSystem.File.ReadAllText(fileUri.LocalPath));
-            }
-            catch (Exception exception)
-            {
-                // I/O classes typically throw a large variety of exceptions
-                // instead of handling each one separately let's just trust the message we get
-                return new(x => x.ErrorOccurredReadingFile(exception.Message));
             }
+
+            return new(new FileWithEncoding(sb.ToString(), sr.CurrentEncoding));
```

Comment:
I am not sure if this is true but I think maybe there is a simpler equivalent implementation. With that said, perhaps I am missing a nuance difference, can you share your thoughts? 

```suggestion
public static ResultWithDiagnostic<FileWithEncoding> ReadWithEncoding(BinaryData data, Encoding fileEncoding, int maxCharacters, Uri fileUri)
{
    var contents = data.ToString();
    if(contents.Length > maxCharacters){
        return new(x => x.FileExceedsMaximumSize(fileUri.LocalPath, maxCharacters, "characters"));
    }
    return new(new FileWithEncoding(contents, sr.CurrentEncoding));
}
```

Reply:
There's no one way to convert a byte stream into a string - there are various possible encodings, which have different binary representations. My understanding is that the `StreamReader` class (`sr` variable) allows the user to pass in an expected encoding, and also does its own encoding detection, even if the user passes in a custom encoding.

We'd have to look at the implementation of `.ToString()`. I didn't do much digging here (just did the minimum to preserve existing behavior), but it's not immediately clear to me how we'd achieve equivalent functionality.

The baselines here exercise some of this functionality if you're interested in understanding more:
https://github.com/Azure/bicep/tree/main/src/Bicep.Core.Samples/Files/baselines/LoadFunctions_CRLF
https://github.com/Azure/bicep/tree/main/src/Bicep.Core.Samples/Files/baselines/InvalidLoadFunctions_CRLF

Reply:
From: https://learn.microsoft.com/en-us/dotnet/api/system.binarydata.tostring?view=dotnet-plat-ext-8.0
> No special treatment is given to the contents of the data. It is merely decoded as a UTF-8 string. For a JPEG or other binary file format, the string will largely be nonsense with many embedded NUL characters. UTF-8 JSON values will look like their file or network representation, including starting and stopping quotes on a string.

So likely using ToString() would not be equivalent since not all encoding types are supported. Thanks for taking the time to reply 👍